### PR TITLE
Join-dependent filter derivation

### PR DESCRIPTION
**Pull request by lnkuiper**
This PR implements a basic version of join-dependent filter derivation as proposed [here](https://homepages.cwi.nl/~boncz/snb-challenge/chokepoints-tpctc.pdf).

This is kind of a niche optimization, which inspects filters like this:
```sql
SELECT *
FROM nation n1
JOIN nation n2
ON ((n1.n_name = 'FRANCE'
        AND n2.n_name = 'GERMANY')
    OR (n1.n_name = 'GERMANY'
        AND n2.n_name = 'FRANCE'));
```
The join filter as a whole cannot be pushed down, because it references tables from both sides. However, we can derive from this two filters that can be pushed down, namely:
```sql
WHERE (n1.n_name = 'FRANCE' OR n1.n_name = 'GERMANY')
  AND (n2.n_name = 'GERMANY' OR n2.n_name = 'FRANCE')
```
By adding this filter, we can greatly reduce both sides of the join before performing the join on the original condition.
This reduces memory pressure, and speeds up TPC-H Q7 significantly.

Regression test plan cost will fail until [this PR](https://github.com/duckdb/duckdb/pull/11129) is merged, as it fixes a bug.

### CODE REVIEW

**Comment by lnkuiper** [comment]
Just realized that this still needs some work

**Comment by lnkuiper** [comment]
I've made the rule as flexible as I think it can be, so it applies to more complex filters now. It's also tested better

**Review by Tmonster — APPROVED** [review]
LGTM!

There is a different way to expand the statements and maintain, but I think it's best to only do it when table filters can be derived. Otherwise we might be expanding/simplifying these conjunction statements for a long time

File: src/optimizer/rule/join_dependent_filter.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+static inline bool ExpressionReferencesMultipleTables(const Expression &binding) {
+	unordered_set<idx_t> table_idxs;
+	GetTableIndices(binding, table_idxs);
+	return table_idxs.size() > 1;
+}
+
+static inline void ExtractConjunctedExpressions(Expression &expression,
+                                                unordered_map<idx_t, unique_ptr<Expression>> &expressions) {
+	if (expression.GetExpressionClass() == ExpressionClass::BOUND_CONJUNCTION &&
```
Comment (Mytherin) [inline src/optimizer/rule/join_dependent_filter.cpp:38]: nit, but this is not really necessary - `ExpressionType::CONJUNCTION_AND` implies `ExpressionClass::BOUND_CONJUNCTION`

File: src/optimizer/rule/join_dependent_filter.cpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	} else {
+		unordered_set<idx_t> table_idxs;
+		GetTableIndices(expression, table_idxs);
+		if (table_idxs.size() != 1) {
+			return; // Needs to reference exactly one table
+		}
+
+		// If there was already an expression, AND it together
+		auto &table_expressions = expressions[*table_idxs.begin()];
+		table_expressions = table_expressions
```
Comment (Mytherin) [inline src/optimizer/rule/join_dependent_filter.cpp:53]: We should probably verify there are no volatile expressions being used, e.g. what happens if we have `x >= random()` here and we copy the `random` expression?

File: src/include/duckdb/optimizer/rule/join_dependent_filter.hpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/optimizer/rule.hpp"
+
+namespace duckdb {
+
+//! Join-dependent filters cannot be pushed down, but we can sometimes derive a filter from them that can
```
Comment (Mytherin) [inline src/include/duckdb/optimizer/rule/join_dependent_filter.hpp:15]: Could we expand the description here to contain the description of the PR? It is not very clear what this rule does by itself.

**Review by Mytherin — COMMENTED** [review]
Thanks! Looks good - some minor comments:

File: src/optimizer/rule/join_dependent_filter.cpp (hunk 2 above)
Comment (lnkuiper) [reply src/optimizer/rule/join_dependent_filter.cpp:52]: Good catch, I've added a check for volatility, and a test for it as well.

**Comment by lnkuiper** [comment]
This optimizer rule actually triggered on the new test `test/sql/copy/csv/recursive_query_csv.test_slow`!

I've had to add an `ORDER BY ALL` to the test.

**Comment by lnkuiper** [comment]
I think this PR is ready to go!

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**